pkg/pillar/hypervisor/kvm.go

```diff
@@ -4,38 +4,383 @@
+  value = "delay"
+
+[global]
+  driver = "ICH9-LPC"
+  property = "disable_s3"
+  value = "1"
+
+[global]
+  driver = "ICH9-LPC"
+  property = "disable_s4"
+  value = "1"
+
+[realtime]
+  mlock = "off"
+
+[rtc]
+  base = "localtime"
+  driftfix = "slew"
+
+[chardev "charmonitor"]
+  backend = "socket"
+  path = "` + kvmStateDir + `{{.DisplayName}}/qmp"
+  server = "on"
+  wait = "off"
+
+[mon "monitor"]
+  chardev = "charmonitor"
+  mode = "control"
+
+[memory]
+  size = "{{.Memory}}"
+
+[smp-opts]
+  cpus = "{{.VCpus}}"
+  sockets = "{{.VCpus}}"
+  cores = "1"
+  threads = "1"
+
+[chardev "charserial0"]
+  backend = "pty"
+
+[device "serial0"]
+  driver = "isa-serial"
+  chardev = "charserial0"
+
+{{if .EnableVnc}}
+[vnc "default"]
+  vnc = "0.0.0.0:{{if .VncDisplay}}{{.VncDisplay}}{{else}}0{{end}}"
+  to = "99"
+{{- if .VncPasswd}}
+  password = "on"
+{{- end -}}
+{{end}}
+
+#[device "video0"]
+#  driver = "qxl-vga"
+#  ram_size = "67108864"
+#  vram_size = "67108864"
+#  vram64_size_mb = "0"
+#  vgamem_mb = "16"
+#  max_outputs = "1"
+#  bus = "pcie.0"
+#  addr = "0x1"
+[device "video0"]
+  driver = "cirrus-vga"
+  vgamem_mb = "16"
+  bus = "pcie.0"
+  addr = "0x1"
+
+[device "pci.1"]
+  driver = "pcie-root-port"
+  port = "0x11"
+  chassis = "2"
+  bus = "pcie.0"
+  addr = "0x2.0x1"
+
+[device "usb"]
+  driver = "qemu-xhci"
+  p2 = "15"
+  p3 = "15"
+  bus = "pci.1"
+  addr = "0x0"
+
+[device "input0"]
+  driver = "usb-tablet"
+  bus = "usb.0"
+  port = "1"
+`
+
+const qemuDiskTemplate = `
+[device "pci.{{.PCIId}}"]
+  driver = "pcie-root-port"
+  port = "0x12"
+  chassis = "3"
+  bus = "pcie.0"
+  addr = "0x2.0x2"
+
+[drive "drive-virtio-disk{{.DiskID}}"]
+  file = "{{.DiskFile}}"
+  format = "{{.DiskFormat}}"
+  if = "none"
+
+[device "virtio-disk{{.DiskID}}"]
+  driver = "virtio-blk-pci"
+  scsi = "off"
+  bus = "pci.{{.PCIId}}"
+  addr = "0x0"
+  drive = "drive-virtio-disk{{.DiskID}}"
+  bootindex = "{{.DiskID}}"
+`
+
+const qemu9PTemplate = `
+# [device "pci.{{.PCIId}}"]
+#  driver = "pcie-root-port"
+#  port = "0x12"
+#  chassis = "3"
+#  bus = "pcie.0"
+#  addr = "0x2.0x2"
+
+[fsdev "fsdev{{.DiskID}}"]
+  fsdriver = "local"
+  security_model = "none"
+  path = "{{.DiskFile}}"
+
+[device "fs{{.DiskID}}"]
+  driver = "virtio-9p-pci"
+  fsdev = "fsdev{{.DiskID}}"
+  mount_tag = "hostshare"
+
+#  bus = "pci.{{.PCIId}}"
+#  addr = "0x0"
+`
+
+const qemuNetTemplate = `
+[device "pci.{{.PCIId}}"]
+  driver = "pcie-root-port"
+  port = "0x10"
+  chassis = "1"
+  bus = "pcie.0"
+  multifunction = "on"
+  addr = "0x2"
+
+[netdev "hostnet{{.NetID}}"]
+  type = "tap"
+  br = "{{.Bridge}}"
+  script = "no"
+  downscript = "no"
+
+[device "net{{.NetID}}"]
+  driver = "e1000e"
+  netdev = "hostnet{{.NetID}}"
+  mac = "{{.Mac}}"
+  bus = "pci.{{.PCIId}}"
+  addr = "0x0"
+`
+
+const kvmStateDir = "/var/run/hypervisor/kvm/"
+
+func logError(format string, a ...interface{}) error {
+	log.Errorf(format, a...)
+	return fmt.Errorf(format, a...)
```

Comment:
Don't need such repeat of the formating. This works just fine and doesn't introduce a new function I need to understand.
err := fmt.Errorf(...)
log.Error(err)
return err

Reply:
I totally get what you're saying @eriknordmark but I have to plea on this one.

Go is super verbose as it is (with all the if err != null and all that) which makes it pretty tedious to actually do the right thing (like proper error reporting and logging). So while you're absolutely correct that we can spell everything in 3 lines, I just feel that condensing it to a single line make is less tedious to type and actually much more readable for me.

Now, I would totally agree that me having this weird little function is not the right way to go about making go less tedious, but how about we promote this little weird function to some kind of a package that starts encapsulating our preferences for doing proper logging and error reporting?

What do others think (I'm special curious to hear from @deitch 'cuz I'm sure he's seen way more idiomatic Go).

Reply:
> Go is super verbose as it is (with all the if err != null and all that)

someone has been reading Bryan's posts on rust. :-)

There isn't a sane way of doing this. The usual method is to have very focused funcs that do not log, only return, so something like:

```go
if err := doSomething(); err != nil {
    return fmt.Errorf("we failed: %v", err)
}
```

and then only do logging at the highest level. Most of the time it makes sense, especially since lower-level functions shouldn't be dealing with file handles (stdin/stdout/stderr), and only the highest level. Of course, that doesn't always work; sometimes you need to log at a lower-level *and* pass something up. In that case, I always do what @eriknordmark suggested:

```go
if err := doSomething(); err != nil {
    msg := fmt.Errorf("we failed: %v", err)
    log.Error(err)
    return err
}
```

The bigger question is, should we be logging from the `hypervisor` package? Or only at the higher level? I did a quick search for `logError` in this PR, I see it in the definition plus another 15 times: 4 in `CreateDomConfig()`, 5 in `Create()`, 1 in `Stop()`, 2 in `Delete()`, 3 in `GetDomCPUs()`.

Why would we log from here? Why not return the error and let the consuming package (likely `domainmgr`) handle logging. `domainmgr` is a `cmd`, so it is "highest-level" and makes sense that it should decide when to log and when not. I get that this may not be realistic, though.

Reply:
Yeah loggin in leaf routines in hypervisor might not make sense. Caller can do that.

Reply:
Fair enough,  @eriknordmark @deitch - tell you what: let me merge this as-is so that @cshari-zededa can make progress, kick off a thread on logging guidelines on eve-tsc@ and depending on the resolution of that thread I promise to update the code one way or the other.